aws/src/main/java/org/apache/iceberg/aws/RESTSigV4Signer.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.aws;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.hc.core5.http.EntityDetails;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.HttpRequest;
+import org.apache.hc.core5.http.HttpRequestInterceptor;
+import org.apache.hc.core5.http.io.entity.StringEntity;
+import org.apache.hc.core5.http.protocol.HttpContext;
+import org.apache.iceberg.exceptions.RESTException;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.signer.Aws4Signer;
+import software.amazon.awssdk.auth.signer.internal.SignerConstant;
+import software.amazon.awssdk.auth.signer.params.Aws4SignerParams;
+import software.amazon.awssdk.auth.signer.params.SignerChecksumParams;
+import software.amazon.awssdk.core.checksums.Algorithm;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpMethod;
+import software.amazon.awssdk.regions.Region;
+
+public class RESTSigV4Signer implements HttpRequestInterceptor {
+
+  private final Aws4Signer signer = Aws4Signer.create();
+  private AwsCredentialsProvider credentialsProvider;
+
+  private String signingName;
+  private Region signingRegion;
+
+  public void initialize(Map<String, String> properties) {
+    AwsProperties awsProperties = new AwsProperties(properties);
+
+    this.signingRegion = awsProperties.restSigv4SigningRegion();
+    this.signingName = awsProperties.restSigv4SigningName();
+    this.credentialsProvider = awsProperties.restSigv4CredentialsProvider();
+  }
+
+  @Override
+  public void process(HttpRequest request, EntityDetails entity, HttpContext context) {
+    try {
+      Aws4SignerParams params =
+          Aws4SignerParams.builder()
+              .signingName(signingName)
+              .signingRegion(signingRegion)
+              .awsCredentials(credentialsProvider.resolveCredentials())
+              .checksumParams(
+                  SignerChecksumParams.builder()
```

Comment:
@jackye1995 it appears the SDK is producing an incorrect content hash for requests without a body.  Based on the docs, SigV4 should produce `e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855` as the body hash for empty requests (GET, HEAD).  However, the actual value is `47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=` which appears to be a base64 encoded representation.

We can work around this by either omitting the hash in those cases, which seems to work, or overriding the produce value, but it seams like we should be able to use the SDK like this by default.

Any thoughts on this?

Reply:
The content hash appears to be [added here](https://github.com/aws/aws-sdk-java-v2/blob/master/core/auth/src/main/java/software/amazon/awssdk/auth/signer/internal/AbstractAws4Signer.java#L452), but this looks like the wrong encoding [based on the docs](https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html)

Reply:
I filed [an issue](https://github.com/aws/aws-sdk-java-v2/issues/3807) with SDK maintainers to see what they they say, but added a workaround to correctly produce the base16 encoded empty body sha256 value.